src/Symfony/Bridge/Doctrine/Form/ChoiceList/EntityChoiceList.php

```diff
@@ -107,16 +114,21 @@ public function __construct(ObjectManager $manager, $class, $labelPath = null, E
         $this->class = $this->classMetadata->getName();
         $this->loaded = is_array($entities) || $entities instanceof \Traversable;
         $this->preferredEntities = $preferredEntities;
+        list(
+            $this->idField,
+            $this->idAsIndex,
+            $this->idAsValue
+        ) = $this->getIdentifierInfoForClass($this->classMetadata);
+
+        if (null !== $this->idField && $this->classMetadata->hasAssociation($this->idField)) {
+            $this->idClassMetadata = $this->em->getClassMetadata(
+                $this->classMetadata->getAssociationTargetClass($this->idField)
+            );
 
-        $identifier = $this->classMetadata->getIdentifierFieldNames();
-
-        if (1 === count($identifier)) {
-            $this->idField = $identifier[0];
-            $this->idAsValue = true;
-
-            if (in_array($this->classMetadata->getTypeOfField($this->idField), array('integer', 'smallint', 'bigint'))) {
-                $this->idAsIndex = true;
-            }
+            list(
+                $this->idAsIndex,
+                $this->idAsValue
+            ) = array_slice($this->getIdentifierInfoForClass($this->idClassMetadata), 1);
```

Comment:
I think you can paste the same code here as above instead of calling `array_slice()`, no? The ID field should remain the same.

For brevity, I still prefer a `initIdentifierInfoForClass()` method.